### ISSUE TITLE
Registers get stuck reading/storing bits

### ISSUE THREAD

**Issue by coreyeng** [Origen-SDK/origen#280]
Hello, I have a problem with a register getting stuck storing bits over and over. This is really easy to miss, because the only way to really see it is to use <code>read!</code> and <code>store!</code> back to back for the <u>the same object<u> (bit collection, register, etc.).

What's happening is when <code>store!</code> or <code>read!</code> is used, arm_debug is moving the data into its own registers. [I think this is an example of that happening.](https://github.com/Origen-SDK/origen_arm_debug/blob/4bff546fbde453882bf6f014c3462930b007b6f6/lib/origen_arm_debug/jtag_dp_controller.rb#L83-L99) Then, this register gets fed into <code>origen_jtag</code> (or <code>origen_swd</code>, in a similar fashion), so that when the JTAG driver tries to [clear the flags](https://github.com/Origen-SDK/origen_jtag/blob/master/lib/origen_jtag/driver.rb#L261), its clearing the flags of an ARM debug register, not the register I gave it.

An example pattern to demonstrate:

~~~
  dut.mem(0x2000_0004).write!(0xDDDD_0000)
  dut.mem(0x2000_0004).read!(0xDDDD_0000)
  dut.mem(0x2000_0004).store!

  dut.mem(0x2000_0004).write!(0xBADC0DE)
  dut.mem(0x2000_0004).read!(0xBADC0DE)
  dut.mem(0x2000_0004).store!

  dut.mem(0x2000_0004).write!(0xAAAA_0000)
  dut.mem(0x2000_0004).read!(0xAAAA_0000)
  dut.mem(0x2000_0004).store!
~~~

Because the flags are never being cleared in my <code>dut.mem(0x2000_0004)</code> register, and because [JTAG looks for stores before reads](https://github.com/Origen-SDK/origen_jtag/blob/master/lib/origen_jtag/driver.rb#L168-L173), I get 1 32-bit assert, and 5 32-bit stores. So what this looks like in the pattern is:

~~~
  dut.mem(0x2000_0004).write!(0xDDDD_0000)
  dut.mem(0x2000_0004).read!(0xDDDD_0000) #=> sets READ flag -> fine
  dut.mem(0x2000_0004).store! #=> sets STORE flag, with READ flag set -> fine

  dut.mem(0x2000_0004).write!(0xBADC0DE)
  dut.mem(0x2000_0004).read!(0xBADC0DE) #=> both READ and STORE are still set -> undesired store occurs, desired read does not occur.
  dut.mem(0x2000_0004).store! #=> both READ and STORE are still set -> desired store occurs, 1 offset from where it want it to be.

  dut.mem(0x2000_0004).write!(0xAAAA_0000)
  dut.mem(0x2000_0004).read!(0xAAAA_0000) #=> another store occurs, read does not.
  dut.mem(0x2000_0004).store! #=> store occurs, now offset by 2 from where I want it to be.
~~~

The solution here looks easy: just have <code>origen_arm_debug</code> clear the flags of the user's register. But, I'm wondering if Origen itself should actually be clearing the flags. When I use <code>read!</code> or <code>store!</code>, I'm expecting only a single read to occur from Origen. The fact that ARM debug actually does a few register writes then a read to accomplish this are details I don't think Origen should care about. Therefore, I'd propose we change the <code>bit_collection</code> to handle clearing the flags after it tells whatever driver its using to read or store. Essentially, I don't think an error in the driver should cause Origen to get stuck reading/storing bits. Also, I don't think its documented anywhere that its the driver's responsibility to clear the flags, so I nominate Origen to do it.

This is also in line with how <code>pins</code> work. Since there's no pin "drivers", like there are register "drivers", Origen handles setting and clearing the state appropriately.

Any thoughts on this? I have no problem implementing and checking this out. I just want to make sure we're all on board with this being Origen's job, not the register drivers'.

**Comment by ginty** [comment]
The ARM debug driver should be clearing these bits, so it is a bug in that driver by current conventions at least.

I have thought about making this change in the past too, but held back because I was nervous about any application code which relied on the current behavior, though I can't really think of any particular scenario.

I also think that for 99% of users this is never an issue, it is really only a concern for someone writing a new driver, but I agree that we could go a step further and remove it as a concern for that subset of users as well.

Speaking for the applications in my domain, I would be OK with making this change.

**Comment by pderouen** [comment]
I've noticed this behavior as well.  To get around it I'm just in the habit of clearing flags in my DUT read/write register methods.  I had just assumed that drivers were expected to leave the flags alone and let the app decide what to do.  That assumption was based on the fact that the overlay flag persists.

Maybe the best place to change this through origen is by wrapping the reg read/write method and add a "clear flags on read/write" that defaults to the current behavior.

**Comment by ginty** [comment]
Drivers not clearing overlays is intentional, but they should be clearing read/store, by calling reg.clear_flags on the original register passed in from the application.

The reasoning behind that is that read/store operations are almost always a one time event, i.e. just because you store the value of a register now doesn't meant that you want to store it the next time you read it.

In the case where a register is overlaid with some data though, then normally you do want to reference that variable overlay data everytime you access that register in a pattern. So for that reason overlays are sticky and are not cleared by drivers.
The application can of course clear it if does not want this default behavior for some reason.

**Comment by pderouen** [comment]
I understand the reasoning behind making overlay sticky.  It just seems cleaner to let reg.read! send the reg to the dut read method and then clear flags when done.  Otherwise every driver will have to be updated.  Do any clear flags?

**Comment by chrisnappi** [comment]
Just chiming in on another place this causes issues (I am somewhat indifferent to where it is fixed) - when we do something like this:

```
  dut.module.reg(:myreg).read! do |r|
      r.bits(:mybit1).read(0x1)
  end

  dut.module.reg(:myreg).read! do |r|
      r.bits(:mybit2).read(0x1)
  end
```
The second read will read BOTH mybit1 and mybit2.  This is obviously not what the code intends to do.  AFAIK all the drivers have this issue currently.

**Comment by ginty** [comment]
> Do any clear flags?

@pderouen, depends which area you work in I guess, I thought they all did until now :-)

On reflection, I guess I mainly use raw JTAG, which does: https://github.com/Origen-SDK/origen_jtag/blob/dfbe54833013b28f13c5f43dea517410d1f1de60/lib/origen_jtag/driver.rb#L261

or else our own custom protocol drivers which also do.

Anyway, given that this seems to be a wide problem I think we should go with Corey's original proposal, i.e. clear them here: https://github.com/Origen-SDK/origen/blob/master/lib/origen/registers/bit_collection.rb#L544

**Comment by pderouen** [comment]
After sleeping on it, I've come full circle and wonder if it wouldn't be better to have the drivers be responsible for clearing the flags after all.  Otherwise, you could get different behavior when read! isn't used to perform the operation.

**Comment by ginty** [comment]
Good point @pderouen, and actually now that I think about it some more, I think it is more of a driver responsibility than Origen core.
Clearing the flags is an acknowledgement to higher up the chain that somebody has performed the read/store as requested and its only really the driver that knows when/if this has been done.

**Comment by chrisnappi** [comment]
Ok @ginty @pderouen , I like the idea that a driver should signal that the action has been completed, but what should be done to models afterwards should be common.  

Right now we may only have this case of flags, but at some point we really need to implement the additional register behaviors(access codes) and maybe other model updates.  

 I think it will be more future proof to have the driver signal an action type is complete, but centralize what should happen after.

**Comment by ginty** [comment]
Sounds good @chrisnappi , so maybe we should take the opportunity to transition to two new bitcollection methods here, `write_complete` and `read_complete`.

'write_complete` would probably be empty right now and do nothing, `read_complete` would internally call `clear_flags`.

**Comment by coreyeng** [comment]
Do we have other methods to initiate a read or store other than read! and store!? I guess I'm still not seeing the benefit to having Origen acknowledge the driver completed the operation. If I use <code>read!/store!</code>, I'm expecting to get vectors for a read or store, or a Ruby exception. To me, Origen's acknowledgement that the read completed is that <code>read_register</code> was called without raising an exception.

Maybe a better wording: is there any case where we would run a read! or store! and not expect those flags to be cleared? We don't necessarily need read! to clear all the flags, maybe just the read flag, but it seems that if every driver should clear the flags, we should factor that out and have Origen do it. We can do the read/write/store complete like @ginty suggested, but I still think Origen should call those, then the driver can override them if it needs to (I don't know why it would need to).

**Comment by ginty** [comment]
The case raise by @pderouen is that it is perfectly legal to bypass Origen's `read!` API if you want to:

~~~ruby
my_reg.my_bits.read(0x5)
# For some valid reason, I want to directly specify the driver this goes to:
dut.some_driver.read_register(my_reg)
~~~

Origen doesn't even know the read happened in that case unless the driver tells it it did.

**Comment by pderouen** [comment]
We do have 1 IP that has registers that have a special interface that uses a different driver than the rest of the dut.  The owner is using direct calls to a driver to make the patterns.

**Comment by coreyeng** [comment]
@ginty, Thanks for the example! That helps clear it up for me. We do have something like that, but we have all of this select-able from inside the read_register method at the application level. Though, during a debug effort for example, I could see using read/stores at the driver level and getting the same issues again.

But going back, I see the read!/store! as Origen interface methods where it should work as best as it can given the driver. So, again assuming that we only ever want read! and store! to do a single operation, I still think those methods should clear the flags since there's no case where they wouldn't, or at the very least raise an exception if the driver doesn't clear them. It seems irresponsible on Origen's part to let these error through if it could check for it. Actually, as I'm writing this, I think failing in Origen's interface would be best as it forces the user to review their driver implementation.

Maybe a larger issue I'm still not seeing is where the line should be drawn between Origen's register interface and the driver and how we can confirm (or help confirm, if not all the way) that a driver is compliant.

**Comment by ginty** [comment]
Some arguments against having Origen do it centrally:

- It is not really its responsibility, so by stepping in it can temporarily mask bad drivers only to have the problem re-emerge later when it is hard for a user to understand/debug - why does it clear for some reads and not others?!
- Clearing (or checking) is a reasonably expensive operations since it involves looping the bits, we shouldn't do it twice on every reg transaction unless it is adding real value

Some proposals:

If we move to a more generic `read_done`, `write_done` kind of interface, then we could have Origen check on the drivers in a more efficient way.

~~~ruby
# In bitcollection when it is sending registers for read/write
@reg.read_completed = false  # Clear this, expecting the driver to set it to give us a quick check
@reg.request(:read_register, options)
unless @reg.read_completed do
  Origen.log.warn "Your read_register driver has not acknowledge the read properly, unexpected behavior could occur"
end
~~~

~~~ruby
# New method added to bit collection to be called by drivers
def read_done
  clear_flags
  @reg.read_completed = true
end
~~~

How do we make drivers correctly use this acknowledge interface in the first place?

Our documentation could/should be better, I would consider moving this section to its own page within the Pattern section - http://origen-sdk.org/origen/guides/pattern/registers/#Writing_a_Driver
Then we can add to it and specifically document this handshake interface.

I would like the new_app_generators to provide a serial and parallel driver plugin template in its test engineering section.

That would contain a skeleton read_register method, with comments on where to add the driver-specific stuff, and it would contain the call to `read_done` at the end with a comment that it is required.

It could also contain a basic test DUT setup and a test which calls dut.some_reg.read! and then tests that the read flags are cleared afterwards.

**Comment by pderouen** [comment]
I like the idea of having the expectations of the driver clearly spelled out.  I like the idea of the skeleton generator and behavior checker even better.

**Comment by coreyeng** [comment]
Okay. Thanks for the discussion! I'll update the arm debug driver to be compliant with what Origen is expecting. I'll open another issue too to track the driver documentation, behavioral checking, and skeleton generator, just to further register it as a 'known shortcoming' (or whatever you want to call it) since I'm not expecting us to get to that soon.

I'll close this one once I release an arm debug driver update.

Thanks!

**Comment by priyavadan** [comment]
@redxeth I believe this ties into the issue you were seeing as well!

**Comment by coreyeng** [comment]
Fixed in [1.1.3](https://origen-sdk.org/arm_debug/release_notes/#v1_1_3)
